modules/twinkleblock.js

```diff
@@ -0,0 +1,870 @@
+      { label: 'Spam/advertising-only account', value: 'uw-soablock' },
+      { label: 'Unapproved bot', value: 'uw-botblock' },
+      { label: 'Violating the three-revert rule', value: 'uw-3block' }
+    ]
+  },
+  {
+    label: 'Username violations',
+    list: [
+      { label: 'Bot username', value: 'uw-botublock' },
+      { label: 'Memorial username soft block', value: 'uw-memorialblock' },
+      { label: 'Promotional username, hard block', value: 'uw-spamublock' },
+      { label: 'Promotional username, soft block', value: 'uw-softerblock' },
+      { label: 'Similar username soft block', value: 'uw-ublock-double' },
+      { label: 'Username violation, soft block', value: 'uw-ublock' },
+      { label: 'Username violation, hard block', value: 'uw-uhblock' },
+      { label: 'Username impersonation hard block', value: 'uw-uhblock-double' },
+      { label: 'Username represents a famous person, soft block', value: 'uw-ublock-famous' },
+      { label: 'Username represents a non-profit, soft block', value: 'uw-causeblock' },
+      { label: 'Username violation, vandalism-only account', value: 'uw-vaublock' }
+    ]
+  },
+  {
+    label: 'Templated reasons',
+    list: [
+      { label: 'anonblock', value: 'anonblock' },
+      { label: 'anonblock - likely a school', value: 'anonblock - school' },
+      { label: 'blocked proxy', value: 'blocked proxy' },
+      { label: 'CheckUser block', value: 'CheckUser block' },
+      { label: 'checkuserblock-account', value: 'checkuserblock-account' },
+      { label: 'checkuserblock-wide', value: 'checkuserblock-wide' },
+      { label: 'colocationwebhost', value: 'colocationwebhost' },
+      { label: 'oversightblock', value: 'oversightblock' },
+      { label: 'school block', value: 'school block' },
+      // { label: 'rangeblock', value: 'rangeblock' },
+      { label: 'tor', value: 'tor' },
+      { label: 'webhostblock', value: 'webhostblock' }
+    ]
+  }
+];
+
+Twinkle.block.callback.filtered_block_groups = function twinkleblockCallbackFilteredBlockGroups(show_template) {
+  return $.map(Twinkle.block.blockGroups, function(value) {
+    var list = $.map(value.list, function(v) {
+        var blockSettings = Twinkle.block.blockPresetsInfo[v.value];
+        var registrationRestrict = blockSettings.forRegisteredOnly ? Twinkle.block.isRegistered : (blockSettings.forAnonOnly ? !Twinkle.block.isRegistered : true);
+        if (!blockSettings.excludeFromTemplates && registrationRestrict) {
+          return {
+            label: (show_template ? '{{' + v.value + '}}: ' : '') + v.label,
+            value: v.value
+          };
+        }
+      });
+    if (list.length) return {
+        label: value.label,
+        list: list
+      };
+  });
+};
+
+Twinkle.block.callback.change_preset = function twinkleblockCallbackChangePreset(e) {
+  var $form = $(e.target.form), key = $form.find('[name=category]').val();
+  if (!key) return;
+
+  $form.find('[name=sub_group]').val(key);
```

Comment:
Is this some sort of hack to "integrate" block and warn? I don't really like it. 

As a general comment, I think creating a messy "web of interdependence" between warn and block is the wrong way to go. Blocks are special-cased in warn half the time anyway; I really think you should just reimplement the warning logic in the block module, to avoid calling back and forth and passing hackish hybrid data objects around.

I'm open to being convinced that the templating logic has enough in common with the other stuff in warn that it can stay there, but you'll have to persuade me.